File: Simulator/src/gjk.cpp
#include "gjk.hpp"
vec2 subtract(vec2 &a, vec2 &b) {
    vec2 sub;
    sub.x = a.x - b.x;
    sub.y = a.y - b.y;
    return sub;
}
vec2 negate(vec2 &v) {
    vec2 v_n;
    v_n.x = -v.x;
    v_n.y = -v.y;
    return v_n;
}
vec2 perpendicular(vec2 &v) {
    vec2 p = {v.y, -v.x};
    return p;
}
double dotProduct(vec2 &a, vec2 &b) {
    return a.x*b.x+a.y*b.y;
}
double lengthSquared(vec2 &v) {
    return v.x*v.x + v.y*v.y;
}
vec2 tripleProduct(vec2 &a, vec2 &b, vec2 &c) {
    vec2 r;
    double ac = a.x*c.x + a.y*c.y;
    double bc = b.x*c.x + b.y*c.y;
    r.x = b.x*ac - a.x*bc;
    r.y = b.y*ac - a.y*bc;
    return r;
}
vec2 averagePoint(std::vector<vec2> &vertices) {
    size_t count = vertices.size();
    vec2 avg = {0., 0.};
    for (size_t i=0; i<count; i++) {
        avg.x += vertices[i].x;
        avg.y += vertices[i].y;
    }
    avg.x /= count;
    avg.y /= count;
    return avg;
}
size_t indexOfFurthestPoint(std::vector<vec2> &vertices, vec2 &d) {
    size_t count = vertices.size();
    double maxProduct = dotProduct(d, vertices[0]);
    size_t index = 0;
    for (size_t i=1; i<count; i++) {
        double product = dotProduct(d, vertices[i]);
        if (product > maxProduct) {
            maxProduct = product;
            index = i;
        }
    }
    return index;
}
vec2 support(std::vector<vec2> &vertices1, std::vector<vec2> &vertices2, vec2 &d) {
    vec2 d_n = negate(d);
    size_t i = indexOfFurthestPoint(vertices1, d);
    size_t j = indexOfFurthestPoint(vertices2, d_n);
    vec2 support = subtract(vertices1[i], vertices2[j]);
    return support;
}
int gjk(std::vector<vec2> &vertices1, std::vector<vec2> &vertices2) {
    size_t index = 0;
    vec2 a, b, c, d, ao, ab, ac, abperp, acperp;
    std::vector<vec2> simplex(3);
    vec2 position1 = averagePoint(vertices1);
    vec2 position2 = averagePoint(vertices2);
    d = subtract(position1, position2);
    if ((d.x == 0) && (d.y == 0)) {
        d.x = 1.;
    }
    a = support(vertices1, vertices2, d);
    simplex[0] = a;
    if (dotProduct(a, d) <= 0) {
        return 0;
    }
    d = negate(a);
    int iter_count = 0;
    while (1) {
        iter_count++;
        a = support(vertices1, vertices2, d);
        simplex[++index] = a;
        if (dotProduct(a, d) <= 0) {
            return 0;
        }
        ao = negate(a);
        if (index < 2) {
            b = simplex[0];
            ab = subtract(b, a);
            d = tripleProduct(ab, ao, ab);
            if (lengthSquared(d) == 0) {
                d = perpendicular(ab);
            }
            continue;
        }
        b = simplex[1];
        c = simplex[0];
        ab = subtract(b, a);
        ac = subtract(c, a);
        acperp = tripleProduct(ab, ac, ac);
        if (dotProduct(acperp, ao) >= 0) {
            d = acperp;
        } else {
            abperp = tripleProduct(ac, ab, ab);
            if (dotProduct(abperp, ao) < 0) {
                return 1;
            }
            simplex[0] = simplex[1];
            d = abperp;
        }
        simplex[1] = simplex[2];
        --index;
    }
    return 0;
}
double Perturbation();
vec2 Jostle(vec2 a);

File: Simulator/python/GJK.py
import numpy as np
import numba

@numba.njit(cache=True)
def perpendicular(pt):
 temp=pt[0]
 pt[0]=pt[1]
 pt[1]=-1*temp
 return pt

@numba.njit(cache=True)
def tripleProduct(a,b,c):
 ac=a.dot(c)
 bc=b.dot(c)
 return b*ac-a*bc

@numba.njit(cache=True)
def avgPoint(vertices):
 return np.sum(vertices,axis=0)/vertices.shape[0]

@numba.njit(cache=True)
def indexOfFurthestPoint(vertices,d):
 return np.argmax(vertices.dot(d))

@numba.njit(cache=True)
def support(vertices1,vertices2,d):
 i=indexOfFurthestPoint(vertices1,d)
 j=indexOfFurthestPoint(vertices2,-d)
 return vertices1[i]-vertices2[j]

@numba.njit(cache=True)
def collision(vertices1,vertices2):
 index=0
 simplex=np.empty((3,2))
 position1=avgPoint(vertices1)
 position2=avgPoint(vertices2)
 d=position1-position2
 if d[0]==0 and d[1]==0:
  d[0]=1.0
 a=support(vertices1,vertices2,d)
 simplex[index,:]=a
 if d.dot(a)<=0:
  return 0
 d=-a
 iter_count=0
 while iter_count<1e3:
  a=support(vertices1,vertices2,d)
  index+=1
  simplex[index,:]=a
  if d.dot(a)<=0:
   return 0
  ao=-a
  if index<2:
   b=simplex[0,:]
   ab=b-a
   d=tripleProduct(ab,ao,ab)
   if np.linalg.norm(d)<1e-10:
    d=perpendicular(ab)
   continue
  b=simplex[1,:]
  c=simplex[0,:]
  ab=b-a
  ac=c-a
  acperp=tripleProduct(ab,ac,ac)
  if acperp.dot(ao)>=0:
   d=acperp
  else:
   abperp=tripleProduct(ac,ab,ab)
   if abperp.dot(ao)<0:
    return 1
   simplex[0,:]=simplex[1,:]
   d=abperp
  simplex[1,:]=simplex[2,:]
  index-=1
  iter_count+=1
 assert(1==0)
 return 0

@numba.njit(cache=True)
def collision_dist(vertices1,vertices2):
 index=0
 simplex=np.empty((3,2))
 position1=avgPoint(vertices1)
 position2=avgPoint(vertices2)
 d=position1-position2
 if d[0]==0 and d[1]==0:
  d[0]=1.0
 a=support(vertices1,vertices2,d)
 simplex[index,:]=a
 if d.dot(a)<=0:
  return distance(vertices1,vertices2,d)
 d=-a
 iter_count=0
 while iter_count<1e3:
  a=support(vertices1,vertices2,d)
  index+=1
  simplex[index,:]=a
  if d.dot(a)<=0:
   return distance(vertices1,vertices2,d)
  ao=-a
  if index<2:
   b=simplex[0,:]
   ab=b-a
   d=tripleProduct(ab,ao,ab)
   if np.linalg.norm(d)<1e-10:
    d=perpendicular(ab)
   continue
  b=simplex[1,:]
  c=simplex[0,:]
  ab=b-a
  ac=c-a
  acperp=tripleProduct(ab,ac,ac)
  if acperp.dot(ao)>=0:
   d=acperp
  else:
   abperp=tripleProduct(ac,ab,ab)
   if abperp.dot(ao)<0:
    return 0.0 
   simplex[0,:]=simplex[1,:]
   d=abperp
  simplex[1,:]=simplex[2,:]
  index-=1
  iter_count+=1
 assert(1==0)
 return distance(vertices1,vertices2,d)

@numba.njit(cache=True)
def distance(vertices1,vertices2,direc):
 a=support(vertices1,vertices2,direc)
 b=support(vertices1,vertices2,-direc)
 d=closestPoint2Origin(a,b)
 dist=np.linalg.norm(d)
 while True:
  if dist<1e-10:
   return dist
  d=-d
  c=support(vertices1,vertices2,d)
  temp1=c.dot(d)
  temp2=a.dot(d)
  if(temp1-temp2)<1e-10:
   return dist
  p1=closestPoint2Origin(a,c)
  p2=closestPoint2Origin(c,b)
  dist1=np.linalg.norm(p1)
  dist2=np.linalg.norm(p2)
  if dist1<dist2:
   b=c
   d=p1
   dist=dist1
  else:
   a=c
   d=p2
   dist=dist2

@numba.njit(cache=True)
def closestPoint2Origin(a,b):
 ab=b-a
 ao=-a
 length=ab.dot(ab)
 if length<1e-10:
  return a
 frac=ao.dot(ab)/length
 if frac<0:
  return a
 if frac>1:
  return b
 return frac*ab+a
# Created by pyminifier (https://github.com/liftoff/pyminifier)
